src/shared/fdset.c

```diff
@@ -126,6 +132,58 @@ int fdset_put_dup(FDSet *s, int fd) {
         return TAKE_FD(copy);
 }
 
+int fdset_put_indexed(FDSet *s, int fd, int index) {
+        int r;
+
+        assert(s);
+        assert(fd >= 0);
+        assert(index >= 0);
+
+        /* Avoid integer overflow in FD_TO_PTR() */
+        if (fd == INT_MAX)
+                return log_debug_errno(SYNTHETIC_ERRNO(EINVAL), "Refusing invalid fd: %d", fd);
+
+        if (hashmap_contains(MAKE_HASHMAP(s), FD_INDEX_TO_PTR(index)))
+                return -EEXIST;
+
+        if (hashmap_contains(MAKE_HASHMAP(s), FD_TO_PTR(fd)))
+                return -EEXIST;
+
+        /* Store both the fd and the index, so that either can be used for lookups. Store the index as
+         * a negative number, so that we know there can be no overlap, since a fd must be >= 0. */
+
+        r = hashmap_put(MAKE_HASHMAP(s), FD_TO_PTR(fd), FD_INDEX_TO_PTR(index));
+        if (r < 0)
+                return r;
+
+        r = hashmap_put(MAKE_HASHMAP(s), FD_INDEX_TO_PTR(index), FD_TO_PTR(fd));
+        if (r < 0) {
+                (void) hashmap_remove(MAKE_HASHMAP(s), FD_TO_PTR(fd));
+                return r;
+        }
+
+        return r;
+}
+
+int fdset_put_dup_indexed(FDSet *s, int fd, int index) {
```

Comment:
hmm, in most cases, wouldn't index set to fdset_size() be what you want?

i mean, sure if you add fds, then remove some earlier ones and add some more you might create an EEXIST case, but we never do that, do we?

Hence, maybe there should be an api to explicitly specify the index, but i think in most cases we probably just could use fdset_size()?

that would make things a ton easier, as the callers would not have to keep count anymore

Reply:
nah, it would be fine to remove from the end, which we even might, due to error paths.

but even if this is misused, then you#d just get a clean EEXIST and hopefuly you learn to do better then.

but the fact is that the fdset is not actually used random access for us. either we just add to it (and possibly drop from the end in some error paths), or we just remove from it. we do not mix things wildly.

Reply:
The problem is not adding duplicates, it's leaving holes around, because then SCM_RIGHTS will fail as it receives an array with EBADF in the middle

Reply:
why would it get that? again, when we put together fdset we never leave holes, we just keep adding.

Reply:
it seems a bit of a footgun to have lying around, but ok, updated